src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java

```diff
@@ -0,0 +1,326 @@
+import net.imglib2.mesh.obj.Triangles;
+import net.imglib2.mesh.obj.Vertices;
+import net.imglib2.mesh.util.MeshUtil;
+import net.imglib2.mesh.util.SortArray;
+import net.imglib2.mesh.util.SortBy;
+import net.imglib2.util.Intervals;
+
+public class InteriorPointTest
+{
+
+	/** Fraction of the specified scale to shift mesh vertices position. */
+	private static final double EPS = 4e-4;
+
+	/** Precision limit for the Moller-Trumbore algorithm. */
+	private static final double EPSILON = 0.0000001;
+
+	/** The X coordinate of the ray vector to cast. We go along positive X. */
+	private static final double RX = 1.;
+
+	/** The Y coordinate of the ray vector to cast. We go along positive X. */
+	private static final double RY = 0.;
+
+	/** The Z coordinate of the ray vector to cast. We go along positive X. */
+	private static final double RZ = 0.;
+
+	/**
+	 * The array of the minimal Z position of all the triangles in the mesh,
+	 * sorted.
+	 */
+	private final double[] minZs;
+
+	/**
+	 * The indices that map the {@link #minZs} value to the triangle index.
+	 */
+	private final int[] indexMin;
+
+	/**
+	 * The array of the maximal Z position of all the triangles in the mesh,
+	 * unsorted.
+	 */
+	private final double[] maxZs;
+
+	/**
+	 * Tmp holder for Z min/max calculations.
+	 */
+	private final int[] indices;
+
+	private final Mesh mesh;
+
+	/** Tmp holder used in the Moller-Trumbore algorithm. */
+	private final double[] tmp = new double[ 3 ];
+
+	/** Holder to store X coords of intersection. */
+	private final TDoubleArrayList xIntersect = new TDoubleArrayList();
+
+	/** Holder to store X normals at the intersection. */
+	private final TDoubleArrayList xNormals = new TDoubleArrayList();
+
+	/**
+	 * Tmp holder to store triangle intersection coordinates in Moller-Trumbore
+	 * algorithm.
+	 */
+	private final double[] intersection;
+
+	private final RealInterval boundingBox;
+
+	public InteriorPointTest( final Mesh mesh, final double scale )
+	{
+		this( mesh, MeshUtil.toRealInterval( Meshes.boundingBox( mesh ) ), scale );
+	}
+
+	public InteriorPointTest( final Mesh mesh, final RealInterval boundingBox, final double scale )
+	{
+		this.mesh = mesh;
+		this.boundingBox = boundingBox;
+
+		/*
+		 * Slice plane Z positions to odd multiples of eps.
+		 */
+		final double eps = EPS * scale;
+
+		// Collect minZ & maxZ of vertices.
+		final Triangles triangles = mesh.triangles();
+		final Vertices vertices = mesh.vertices();
+
+		this.minZs = new double[ ( int ) triangles.size() ];
+		this.maxZs = new double[ ( int ) triangles.size() ];
+		for ( int t = 0; t < triangles.size(); t++ )
+		{
+			final long v0 = triangles.vertex0( t );
+			final long v1 = triangles.vertex1( t );
+			final long v2 = triangles.vertex2( t );
+
+			final double minZ = minZ( vertices, v0, v1, v2, eps );
+			minZs[ t ] = minZ;
+			final double maxZ = maxZ( vertices, v0, v1, v2, eps );
+			maxZs[ t ] = maxZ;
+		}
+		this.indexMin = SortArray.quicksort( minZs );
```

Comment:
It's crazy to me that we have our own quicksort implementation in this library, just to get around issues of sorting indices vs values, and boxing/unboxing of ints. I did some digging to see what libraries exist providing this functionality without a custom sort implementation. E.g. something with an `IntComparator` that has signature `int compareTo(int a, int b)` and then sort routines that operate via such a comparator object. But none of Trove, Guava, or Apache Commons provides this. Only Fastutil, as far as I could find, but the fastutil-core JAR is already >6MB. :confused: 

Is it worth a fastutil dependency to get rid of this custom implementation? At least fastutil doesn't have any dependencies of its own...

Reply:
As above, let's review this after we thinned the dependencies.